### PR TITLE
[mono] Fix StackTrace from a dim and Vtable offsets for static interface method

### LINKED ISSUE #60486: [mono] Generic Default Interface Method crash with Environment.StackTrace
### Description

Calling `Environment.StackTrace` from a default interface method of a generic interface causes a crash in the Mono runtime.

### Reproduction Steps

Compile and run the following program with the Mono runtime...

```
using System;

namespace DefItf
{
    public interface IPublisher<out TData>
    {
        event Action<TData> OnPublish;
    }

    public interface ISubscriber<T>
    {
        void OnReceiveSubscription(T data);

        void Subscribe(IPublisher<T> publisher)
        {
            InternalSubscribe(this, publisher);
        }

        void Unsubscribe(IPublisher<T> publisher)
        {
            InternalUnsubscribe(this, publisher);
        }

        protected static void InternalSubscribe(ISubscriber<T> subscriber, IPublisher<T> publisher)
        {
            publisher.OnPublish += subscriber.OnReceiveSubscription;
            Console.WriteLine(Environment.StackTrace.ToString());
        }

        protected static void InternalUnsubscribe(ISubscriber<T> subscriber, IPublisher<T> publisher)
        {
            publisher.OnPublish -= subscriber.OnReceiveSubscription;
            Console.WriteLine(Environment.StackTrace.ToString());
        }
    }

    public class PubTest :IPublisher<InputData>
    {
        public event Action<InputData> OnPublish;

        public void Call() => OnPublish?.Invoke(new InputData());
    }

    public class InputData
    {
        public int i;
    }

    public class Program : ISubscriber<InputData>
    {
        static void Main(string[] args)
        {
            new Program().Start();
        }
        
        // Start is called before the first frame update
        public void Start()
        {
            var pub = new PubTest();
            var sub = (ISubscriber<InputData>)this;
            sub.Subscribe(pub);
            pub.Call();
            sub.Unsubscribe(pub);
        }

        public void Subscribe(IPublisher<InputData> publisher)
        {
            ISubscriber<InputData>.InternalSubscribe(this, publisher);
        }

        public void OnReceiveSubscription(InputData data)
        {
            Console.WriteLine($"do something");
        }
    }
}
```

### Expected behavior

Program completes without crashing and the output is written to the screen.

### Actual behavior

Program crashes.

### Regression?

Not a regression

### Known Workarounds

No known workarounds.

### Configuration

- Tested with Mono runtime build from code, main branch 56d807d4a012cce1a98cdbea30928b438609c424.
- Verified on Windows and Linux x64

### Other information

_No response_

### PR DESCRIPTION
**Pull request by thaystg**
- Fix StackTrace when called from a DIM.
- Fix the other test case that was added for @bholmes, and this case when the method `TestMethod5` was being called it was executing `TestMethod10`, and this was fixed skipping static interface methods when was calculating vtable offsets.

The fix was completely done by @vargaz, I just opened the PR.
Thanks @vargaz .

Fix https://github.com/dotnet/runtime/issues/60486

### CODE REVIEW

**Comment by bholmes** [comment]
I created a new program with more cases.

https://gist.github.com/bholmes/e3ccaf9d044b7261d6865c8f0173dc64

Method5 appears to be broken but I think this patch is good.

**Review by lambdageek — APPROVED** [review]

File: src/mono/mono/metadata/object.c [hunk 1]
```diff
@@ -1518,6 +1518,8 @@ build_imt_slots (MonoClass *klass, MonoVTable *vt, gpointer* imt, GSList *extra_
 				 * add_imt_builder_entry anyway.
 				 */
 				method = mono_class_get_method_by_index (mono_class_get_generic_class (iface)->container_class, method_slot_in_interface);
+				if (m_method_is_static (method))
+					continue;
```
Comment (lambdageek) [inline src/mono/mono/metadata/object.c:1522]: We're going to need to revisit this (and line 1535, below) when non-abstract static virtual methods are going to be supported

**Comment by bholmes** [comment]
If this is approved, I want this to go back to mono/mono. Let me know if and when you want me to back port these changes to that repo. (assuming that automation will fail.)

**Comment by lambdageek** [comment]
There are CoreCLR and Mono test failures from the new test.
1. CoreCLR failures are because the new test methods should all be marked `NoInlining`
2. Mono failure is because the validation for 
   ```
      t1.TestMethod5(null, new [] {new StackFrame ("TestItf1`1", "TestMethod3"), new StackFrame ("TestItf2`1.TestItf1", "TestMethod5")})
    ```
   is too strict - the actual string in Mono is ``TestItf2`1[[InputData, github60486, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null]].TestItf1<TT>.``   So we should make something like ``new StackFrame (new [] {"TestItf2`1", "TestItf1"}, "TestMethod5")``  so that the class name can be in multiple parts.


In both cases, it's an issue with the testcase, not with the implementation.  (So the mono backport is still ok)

**Review by lambdageek — APPROVED** [review]

**Pull request merged**